File: neetcode/ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isValidBST(TreeNode root) {
        // THOUGHT PROCESS:
        // Use recursion with value bounds to validate the BST property for every node.
        // Each node must lie strictly between the lower and upper bounds passed from ancestors.
        // Time O(n) Space O(h)
        //
        // Example (invalid BST):
        //     5
        //    / \
        //   1   4
        //      / \
        //     3   6
        // Bounds checks:
        //   -inf < 5 < +inf
        //   -inf < 1 < 5
        //   5 < 4 < +inf   <-- violated here

        // Return true only if every node satisfies lower < node.val < upper
        return valid(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean valid(TreeNode node, long lower, long upper) {
        if (node == null) return true;

        // Check that node value is strictly inside (lower, upper)
        if (!(node.val > lower && node.val < upper)) return false;

        // Recurse: left in (lower, node.val) and right in (node.val, upper)
        return valid(node.left, lower, node.val)
            && valid(node.right, node.val, upper);
    }
}

File: meta/trees/validate_binary_search_tree.py
"""
Determine if a binary tree is a valid binary search tree (BST).
"""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from typing import Optional

# Time: O(N) - we visit each node once
# Space: O(H) - recursion stack up to height of tree
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def valid(node: Optional[TreeNode], upper: int, lower: int) -> bool:
            # empty node is valid
            if node is None:
                return True
            # current node must be within bounds
            if not (upper > node.val > lower):
                return False
            # left child's upper bound is current node's value because it must be less
            # right child's lower bound is current node's value because it must be greater
            return valid(node.left, node.val, lower) and valid(node.right, upper, node.val)
        
        # start recursion with infinite bounds
        return valid(root, float('inf'), float('-inf'))